### DISCUSSION TITLE
Read from a csv file

### DISCUSSION THREAD

**Discussion by dm-ibtissam** [exceljs/exceljs#2480] [Q&A]
I have a csv file uploaded in input type file and I want to read this file the exceljs but I don't know how ? can you please help me with that.
In the documentation it's not mention how to read before prepossessing with exceljs this are the examples in the documentation :

```
// read from a file
var workbook = new Excel.Workbook();
workbook.csv.readFile(filename)
    .then(function(worksheet) {
        // use workbook or worksheet
    });
 
// read from a stream
var workbook = new Excel.Workbook();
workbook.csv.read(stream)
    .then(function(worksheet) {
        // use workbook or worksheet
    });
 
// pipe from stream
var workbook = new Excel.Workbook();
stream.pipe(workbook.csv.createInputStream());
```
This the begining of my code :

```
function ValidateFile(input) {
  
    var files = input.files || [];
    if (!files.length) return;
    var file = files[0];

    fileName = file.name;
    
   //what to do here ???

   var workbook = new Excel.Workbook();
   workbook.csv.read(stream)
    .then(function(worksheet) {
        
    });
}
```

Wich one can I use and how ?
Thank you.

**Comment by boyepanthera** [comment]
Are you using it in a NodeJs environment? 

If so, ensure the csv file is uploaded to your server, you can use [`multer`](https://www.npmjs.com/package/multer) npm package to achieve that if getting it from your client.

so let's say you have uploaded the csv file into your server in a folder called upload

You can read the file like this

```
const Excel = require('exceljs')
const filename = 'upload/yourfile.csv'
const workbook =  new Excel.Workbook()

workbook.csv.readFile(filename)
  .then(worksheet => {
    console.log(worksheet.getCell('A1').value)
  });
```
What the code does is to read your csv file and return the sheet1 as worksheet, using .then() promise syntax we use the getCell method on the worksheet to get the value from cell A1 in your csv and print to the console.

**Comment by zhangchenle666** [comment]
`import fs from 'fs'
`import { Duplex } from 'stream'`

`function bufferToStream (buffer) {`
`  let stream = new Duplex()`
`  stream.push(buffer)`
`  stream.push(null)`
`  return stream`
`}`

`fs.readFile(filepath, (err, buffer) => {`
`  if (err) {`
`    console.log(err)`
`  } else {`
`    const stream = bufferToStream (buffer)`
`  workbook.csv.read(stream)`
`  .then(function(worksheet) {`
`    ...`
`  }).catch(err => {`
`    console.log(err)`
`  })`
`}`
`})`

**Comment by Shiko1st** [comment]
Is there a chance to read csv and xlsx in browser?

**Comment by 1whatleytay** [comment]
This is really frustrating. Xlsx works really well in browser via the `load` method if you load an `ArrayBuffer` via `FileReader`, but `load` is inexplicably missing for CSVs. I can't imagine why, a corresponding `writeBuffer` method exists, why not `load`??

The need to load a csv file is always present on browser, even if it is slower or less efficient. It makes absolutely no sense why CSV requires a node environment but Xlsx does not.

**Comment by himmanshuu** [comment]
Is there a workaround for reading a uploaded csv from the browser?

**Comment by jakebott** [comment]
As a workaround I have used the `readable-web-to-node-stream` package, however it would still be very useful to have support for `workbook.csv.load()` as this is a bit messy.

```typescript
import type { Csv } from "exceljs";
import ExcelJS from "exceljs";
import { ReadableWebToNodeStream } from "readable-web-to-node-stream";

const workbook = new ExcelJS.Workbook();
const stream = new ReadableWebToNodeStream(
    file.stream()
) as unknown as Parameters<Csv["read"]>[0];
await workbook.csv.read(stream);
```

**Comment by sushantdhende** [comment]
@jakebott, how to **stream** a csv file if say file name is **userdetails.csv**? (without using **readable-web-to-node-stream**).
I just want to use exceljs to read csv(that has large no. records) as stream & write that stream data to an xlsx file?

**Comment by stichiboi** [comment]
> As a workaround I have used the `readable-web-to-node-stream` package, however it would still be very useful to have support for `workbook.csv.load()` as this is a bit messy.

Not working for me, I get an error

```ReferenceError: process is not defined```

It seems Exceljs is still trying to access Nodejs `Readable`


I ended up using [Papaparse](https://github.com/mholt/PapaParse), and then manually create a worksheet
```typescript

const workbook = new ExcelJS.Workbook();

Papa.parse(file, {
    complete(results: ParseResult<string>) {
        console.log(results);
        const worksheet = workbook.addWorksheet(file.name);
        worksheet.addRows(results.data);
    }
});
```

**Reply by basvanmeurs** [reply]
The error can also be worked around like this:

```typescript
        (window as any).process = {
            nextTick(cb: (...args: any[]) => void, ...args: any[]) {
                setTimeout(() => cb(...args), 0);
            },
        };
        const stream = new ReadableWebToNodeStream(file.stream());
        const workbook = new ExcelJs.Workbook();
        await workbook.csv.read(stream as any);
```

**Comment by Tomasalvarez15** [comment]
Hi everyone!

I had the same problem and I solved it using this trick.

For the xlsx files I was using load like this

```
import { Workbook } from 'exceljs';
const workbook = new Workbook();
const content = await workbook.xlsx.load(file.buffer.data);
const workSheet = content.worksheets[0];
```

And for the csv files I used the Readable and Buffer libraries to read the file buffer. The Buffer library doesn't have to be imported as it is imported already on Node js.

```
import { Workbook } from 'exceljs';
import { Readable } from 'stream';
const workbook = new Workbook();
const workSheet = await workbook.csv.read(
        Readable.from(Buffer.from(file.buffer.data))
      );
```

If I printed the file Object with console.log(file) this would be printed on the console.

```
{
  fieldname: 'file',
  originalname: 'text.csv',
  encoding: '7bit',
  mimetype: 'text/csv',
  buffer: {
    type: 'Buffer',
    data: [
      117, 115, 101, 114, 110,  97, 109, 101,  44,  97, 109, 111,
      117, 110, 116,  44,  97, 115, 115, 101, 116,  13,  10, 114,
      105,  99, 107,  97, 110, 100, 109, 111, 114, 116, 121,  44,
       49,  48,  48,  48,  44,  85,  83,  68,  80,  13,  10, 109,
       97, 115, 111, 110, 109, 111, 117, 110, 116,  44,  53,  48,
       48,  48,  44,  85,  83,  68,  84,  13,  10, 108, 111,  99,
      111, 116, 101, 115, 116, 101, 114,  50,  44,  51,  48,  48,
       48,  44,  85,  83,  68,  84
    ]
  },
  size: 90
}
```

**Comment by huangruichang** [comment]
> As a workaround I have used the `readable-web-to-node-stream` package, however it would still be very useful to have support for `workbook.csv.load()` as this is a bit messy.
> 
> ```ts
> import type { Csv } from "exceljs";
> import ExcelJS from "exceljs";
> import { ReadableWebToNodeStream } from "readable-web-to-node-stream";
> 
> const workbook = new ExcelJS.Workbook();
> const stream = new ReadableWebToNodeStream(
>     file.stream()
> ) as unknown as Parameters<Csv["read"]>[0];
> await workbook.csv.read(stream);
> ```

It works! Perfectly! Don't be confused by that shit 'thumbs down emoji'

**Comment by adrianmiles** [comment]
Yeepers, what a pain, this is how I got it working:
```
  const arrayBuffer = await file.arrayBuffer();

  const readable = new Readable();
  readable._read = () => {}; // _read is required
  readable.push(new Uint8Array(arrayBuffer));
  readable.push(null);

  const worksheet = await workbook.csv.read(readable);

```
